### PR TITLE
Let Tensor comm wait for CUDA ops on new devices in RPC response

### LINKED ISSUE #53287: DISABLED test_device_maps_gpu (__main__.TensorPipeTensorPipeAgentRpcTestWithSpawn)
Not disabling the test yet to collect more instances. 

https://app.circleci.com/pipelines/github/pytorch/pytorch/280939/workflows/2729b18f-c934-47d8-bd37-e9d784abaca4/jobs/11306771/steps

```
Mar 04 14:12:40 ======================================================================
Mar 04 14:12:40 ERROR [5.437s]: test_device_maps_gpu (__main__.TensorPipeTensorPipeAgentRpcTestWithSpawn)
Mar 04 14:12:40 ----------------------------------------------------------------------
Mar 04 14:12:40 Traceback (most recent call last):
Mar 04 14:12:40   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 286, in wrapper
Mar 04 14:12:40     self._join_processes(fn)
Mar 04 14:12:40   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 418, in _join_processes
Mar 04 14:12:40     self._check_return_codes(elapsed_time)
Mar 04 14:12:40   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 461, in _check_return_codes
Mar 04 14:12:40     raise RuntimeError(error)
Mar 04 14:12:40 RuntimeError: Process 3 exited with error code 10 and exception:
Mar 04 14:12:40 Traceback (most recent call last):
Mar 04 14:12:40   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 366, in run_test
Mar 04 14:12:40     getattr(self, test_name)()
Mar 04 14:12:40   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 288, in wrapper
Mar 04 14:12:40     fn()
Mar 04 14:12:40   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 99, in wrapper
Mar 04 14:12:40     return func(*args, **kwargs)
Mar 04 14:12:40   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/distributed/rpc/rpc_test.py", line 4709, in test_device_maps_gpu
Mar 04 14:12:40     self.assertEqual(ret, (torch.zeros(2) + torch.ones(2)).to(1))
Mar 04 14:12:40   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1189, in assertEqual
Mar 04 14:12:40     super().assertTrue(result, msg=self._get_assert_msg(msg, debug_msg=debug_msg))
Mar 04 14:12:40   File "/opt/conda/lib/python3.6/unittest/case.py", line 682, in assertTrue
Mar 04 14:12:40     raise self.failureException(msg)
Mar 04 14:12:40 AssertionError: False is not true : Tensors failed to compare as equal!With rtol=1.3e-06 and atol=1e-05, found 2 element(s) (out of 2) whose difference(s) exceeded the margin of error (including 0 nan comparisons). The greatest difference was 1.0 (0.0 vs. 1.0), which occurred at index 0.
Mar 04 14:12:40 
Mar 04 14:12:40 
Mar 04 14:12:40 
Mar 04 14:12:41 ----------------------------------------------------------------------
```

```
Mar 04 14:10:33   test_device_maps_gpu (__main__.TensorPipeTensorPipeAgentRpcTestWithSpawn) ... ERROR:root:Caught exception: 
Mar 04 14:10:33 Traceback (most recent call last):
Mar 04 14:10:33   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 366, in run_test
Mar 04 14:10:33     getattr(self, test_name)()
Mar 04 14:10:33   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 288, in wrapper
Mar 04 14:10:33     fn()
Mar 04 14:10:33   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 99, in wrapper
Mar 04 14:10:33     return func(*args, **kwargs)
Mar 04 14:10:33   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/distributed/rpc/rpc_test.py", line 4709, in test_device_maps_gpu
Mar 04 14:10:33     self.assertEqual(ret, (torch.zeros(2) + torch.ones(2)).to(1))
Mar 04 14:10:33   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1189, in assertEqual
Mar 04 14:10:33     super().assertTrue(result, msg=self._get_assert_msg(msg, debug_msg=debug_msg))
Mar 04 14:10:33   File "/opt/conda/lib/python3.6/unittest/case.py", line 682, in assertTrue
Mar 04 14:10:33     raise self.failureException(msg)
Mar 04 14:10:33 AssertionError: False is not true : Tensors failed to compare as equal!With rtol=1.3e-06 and atol=1e-05, found 2 element(s) (out of 2) whose difference(s) exceeded the margin of error (including 0 nan comparisons). The greatest difference was 1.0 (0.0 vs. 1.0), which occurred at index 0.
Mar 04 14:10:33 exiting process with exit code: 10
Mar 04 14:10:33 Process 3 terminated with exit code 10, terminating remaining processes.
Mar 04 14:10:33 ERROR (5.437s)
```

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu

### PR DESCRIPTION
**Pull request by mrshenli**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53868 Expose Request Devices to Python
* **#53742 Let Tensor comm wait for CUDA ops on new devices in RPC response**

closes #53287

When using CUDA RPC, if the user function uses a new device which
is not used by any Tensors in the request, current TensorPipeAgent
cannot properly synchronize communication and computation. This is
because the agent cannot predict which devices a user function will
use and is not aware of which stream on that device it should wait
for. Two naive solutions might be 1) create a context/stream for
all available devices, 2) always synchronize with current/default
stream on the new device. However, neither of these would work. The
first one would create CUDA context on a device even if the device
was not used in the RPC. As a result, it would introduce additional
CUDA memory overhead, or worse, lead to deadlock. The second one
will sync all background RPCs to the same current/default stream,
which would interfere with user logic on the user main thread and
contentions on that stream could lead to perfermance problem and
correctness might be hard to argue.

This PR is step 1 towards addressing the problem. We now require
users to synchronize their new device streams to one of the current
stream on existing devices. Then, the agent would let communications
on the new device to wait for streams on existing devices.

After this, we still need to solve the problem of CPU-request +
CUDA-reponse, where we don't have a existing stream at all. One
possibility might be letting users to specify streams for the
user function.

Differential Revision: [D26957172](https://our.internmc.facebook.com/intern/diff/D26957172)

### CODE REVIEW

File: torch/csrc/distributed/rpc/tensorpipe_utils.h [hunk 1]
```diff
@@ -97,15 +99,32 @@ struct TORCH_CUDA_CPP_API CudaLazyStreamContext : public LazyStreamContext {
     if (iter == streams_.end()) {
       auto cudaStream = at::cuda::getStreamFromPool(
           /* isHighPriority */ false, /* device */ index);
+      for (auto& barrierEvent : barrierEvents_) {
+        barrierEvent.block(cudaStream);
```
Comment (mrzzd) [inline torch/csrc/distributed/rpc/tensorpipe_utils.h:103]: I am not sure if I understand this part.  When we reach here, there has been no stream_ for given device  in streams_ array. So all barierEvent are for other gpus. How you can block a steam of one device on events in streams of other devices?

Comment (mrshenli) [reply torch/csrc/distributed/rpc/tensorpipe_utils.h:103]: > How you can block a steam of one device on events in streams of other devices?

IIUC, events on one device can block streams from a different device. Quote from CUDAEvent doc:

> The underlying CUDA events are lazily initialized when the event is first recorded or exported to another process. After creation, only streams on the same device may record the event. However, streams on any device can wait on the event.

File: torch/testing/_internal/distributed/rpc/rpc_test.py [hunk 2]
```diff
@@ -5000,6 +5019,7 @@ def test_device_maps_multi_gpu_self(self):
     @staticmethod
     def _gpu_add_return_to_gpu(x, y):
         if x.device.type == 'cpu' and y.device.type == 'cpu':
+            # FIXME: How do we sync ops on new device with CPU-only requests?
             return (x + y).to(0), (x - y).to(1), (x * y).to(2), (x / y).to(3)
```
Comment (mrshenli) [inline torch/testing/_internal/distributed/rpc/rpc_test.py:5023]: this should be fine today, as CPU->CUDA copy is blocking without pin memory

File: torch/testing/_internal/distributed/rpc/rpc_test.py [hunk 3]
```diff
@@ -5073,6 +5093,7 @@ def test_device_maps_return_to_gpu_self(self):
 
     @staticmethod
     def _add_to_gpu(x, y):
+        # FIXME: How do we sync ops on new device with CPU-only requests?
```
Comment (mrshenli) [inline torch/testing/_internal/distributed/rpc/rpc_test.py:5096]: ditto

File: torch/csrc/distributed/rpc/tensorpipe_utils.h (hunk 1 above)
Comment (lw) [reply torch/csrc/distributed/rpc/tensorpipe_utils.h:103]: I think the way this code ended up being written is very "implicit" and thus hard to understand at first glance. If I got it right, this method is called at two separate stages on the same object (first during allocation of the incoming request, then during serialization of the outgoing response), and in these two occurrences the circumstances are quite different: the first time the barrier events are empty, hence this extra code has no effect, whereas the second time there are barrier events, but this code is only executed for streams that were not already accessed the first time.

Hence only _new_ devices that were added by the callback end up waiting for barrier events, thus overall I think it may work, but it's rather hard to follow as the same code behaves very differently in different cases, based on "distant interactions".

**Review by lw — COMMENTED** [review]
Thanks for investigating that flake and getting to the bottom of this! It's a great catch!

I find it interesting that this issue is, at its core, the same exact one we were facing with the callbacks of CUDAFuture: there as well we're not correctly synchronizing the result (i.e., the child future) with any _additional_ device that is used. I think we should aim at finding a single solution for both issues, to preserve consistency and thus avoid surprising and unintuitive behavior.

For CUDAFuture we don't currently have a full "solution", all we have is an idea for a workaround. Provided we implement the `then_async` method (the name is still up for discussion), a user could do this:
```
def cb(parent_fut):
    t = parent_fut.value()
    assert(t.device() == "cuda:0")
    res = torch.futures.Future()
    stream = torch.cuda.Stream("cuda:1")
    stream.wait_stream(torch.cuda.current_stream("cuda:0")
    with torch.cuda.stream(stream):
        s = t.to("cuda:1")
        res.mark_complete(s)
    return s

child_fut = parent_fut.then_async(cb)
```
I don't know if a similar approach could work for remote RPC functions. In short, we'd need to have a way _other than returning_ for the user to tell us that their result is ready, so that we can capture the current streams that the user might have set before those streams are un-set when the function exits.

While we lack any "correct" way of doing this, a safe option might be to just forbid the user from changing devices in a remote RPC function. As I mentioned elsewhere, I think it's preferable to impose additional restrictions now and possibly lift them later if we find out a good way to do so, as if we did it the other way around we'd risk being stuck with a bad solution due to backwards compatibility. Note that if we decided to introduce this restriction it would be rather easy to introduce a check (i.e., compare the set of devices of a request with the ones of the result) and return an explicit error message.

I also want to reiterate my thinking that if users fully embrace RPC they should never need to manually move tensors to other devices, as any such move could be replaced by a RPC call, thus meaning that each remote RPC function just takes care of processing on a fixed set of devices.

**Comment by mrshenli** [comment]
> I find it interesting that this issue is, at its core, the same exact one we were facing with the callbacks of CUDAFuture: there as well we're not correctly synchronizing the result (i.e., the child future) with any additional device that is used. I think we should aim at finding a single solution for both issues, to preserve consistency and thus avoid surprising and unintuitive behavior.

Yep, I agree that they are similar and it will be great if we could have a consistent solution. On the other hand, there are also differences. For example, for `CUDAFuture.then()`, we could allow users to provide a list of streams and then operate on those streams, and this is also consistent with where cuda-aware c10d APIs are heading (i.e., allow users to configure streams). However, this won't work for RPC, as local stream might not be available on the remote side. 


> For CUDAFuture we don't currently have a full "solution", all we have is an idea for a workaround. Provided we implement the then_async method (the name is still up for discussion), a user could do this:

I didn't get how the example works. Should the `cb` function return `res` `Future` object instead? If so, I guess it can work for RPC, but we will need to modify `@rpc.functions.async_execution` a bit. The downside is it's a little verbose, but a helper should solve that.

File: torch/csrc/distributed/rpc/tensorpipe_utils.h (hunk 1 above)
Comment (mrshenli) [reply torch/csrc/distributed/rpc/tensorpipe_utils.h:103]: > I think the way this code ended up being written is very "implicit" and thus hard to understand at first glance. 

Yep, this is a temporary solution to remove flakiness from master. I would like kicking off these tests again and see if this PR fix the flakiness. I am not confident this is the only problem we have. 

As we discussed offline, I was thinking about a long solution as below:

```python
@rpc.functions.enable_cuda(devices=[0, 1])
def user_func(x):
  assert(x.device = "cuda:0")
  return x.to(1)
```

This API should be sufficient to allow the remote side to synchronize streams correctly. And when this is in, I will remove the current solution in this PR. I didn't go that way initially because I don't wanna introduce API changes before we reach consensus on that. I will the similar solution could also work for `CUDAFuture`?

> I also want to reiterate my thinking that if users fully embrace RPC they should never need to manually move tensors to other devices, as any such move could be replaced by a RPC call, thus meaning that each remote RPC function just takes care of processing on a fixed set of devices.

Yep, we can also do this initially. If we all agree this is better solution for now, I will disable all cross-device tests.

Comment (mrshenli) [reply torch/csrc/distributed/rpc/tensorpipe_utils.h:103]: We can also be explicit on this. If the function returns a tensor on a different device but the function wasn't decorated that way, we just error out.

**Comment by lw** [comment]
The decorator you proposed (i.e., `@rpc.functions.enable_cuda`) could be a solution to this. Although I must say I'm always a bit wary when talking about adding an entirely new API construct, hence personally I'd keep this idea as an option but I'd wait to implement it until users actually demand it and we're sure there's need for it that cannot be addressed in other ways. This is the same approach we took with per-call device maps, in a sense. My fear here is that we might be over-engineering something (with all the long-term maintenance burden that a new API entails) which we're not yet sure is needed in practice.

As for this PR, do you need to land it in order to see if it reduces flakiness in tests? For the same reasons as above, I'd rather hold on changing the API of RPC if the main motivation is our own tests (and not some real user), especially if this change is a stopgap for which we still don't agree on a longterm strategy and which therefore might end up remaining in the codebase for a long time.

**Comment by mrshenli** [comment]
> As for this PR, do you need to land it in order to see if it reduces flakiness in tests? 

We need to do something to remove the flakiness. I see two options in short-term 1) disable all cross-device tests 2) land this PR. Let me know which one you prefer.

**Comment by mrshenli** [comment]
close, in favor of #56405
